src/relay/backend/vm/compiler.cc

```diff
@@ -618,38 +469,120 @@ class VMFunctionCompiler : ExprFunctor<void(const Expr& expr)> {
       op_index = context_->seen_funcs[cfunc->funcs[0]];
     }
 
-    Emit(Instruction::InvokePacked(op_index, arity, return_count, unpacked_arg_regs));
-
-    if (return_count > 1) {
-      // return value is a tuple, we need to create a tuple
-      std::vector<Index> fields_registers;
-      for (size_t i = arity - return_count; i < arity; ++i) {
-        fields_registers.push_back(unpacked_arg_regs[i]);
-      }
-      Emit(Instruction::AllocADT(0, return_count, fields_registers, NewRegister()));
-    }
+    Emit(Instruction::InvokePacked(op_index,
+      argument_registers.size(),
+      output_tuple->fields.size(),
+      argument_registers));
   }
 
   void VisitExpr_(const CallNode* call_node) {
+    Expr op = call_node->op;
+
+    // First we handle the case in which we are using an opaque
+    // operator used to define a sub-dialect, such as memory
+    // allocation operations.
+    if (op.as<OpNode>()) {
+      OpMatch<void> matcher;
+      matcher.Match("memory.invoke_tvm_op",
+        [this](const Array<Expr>& args, const Attrs& attrs, const Array<Type>& type_arg) {
+          CHECK_EQ(args.size(), 3);
+          EmitInvokeTVMOp(Downcast<Function>(args[0]), args[1], args[2]);
+      }).Match("memory.alloc_tensor",
+        [this](const Array<Expr>& args, const Attrs& attrs, const Array<Type>& type_arg) {
+          CHECK_EQ(args.size(), 2);
+
+          // Get the attributes.
+          auto alloc_attrs = attrs.as<AllocTensorAttrs>();
+          CHECK(alloc_attrs != nullptr)
+              << "must be the alloc tensor attrs";
+          auto dtype = alloc_attrs->dtype;
+
+          // The storage will be passed dynamically.
+          this->VisitExpr(args[0]);
+          auto storage_register = last_register_;
+
+          // If the shape is constant then we will emit a static tensor allocation instruction.
+          auto const_shape = args[1].as<ConstantNode>();
+
+          if (const_shape) {
+            NDArray shape = const_shape->data;
+            std::vector<int64_t> raw_shape;
+            DLTensor tensor = shape.ToDLPack()->dl_tensor;
+            CHECK_EQ(tensor.ndim, 1u);
+            CHECK_EQ(tensor.dtype.code, 0U)
+              << "found " << tensor.dtype.code;
+
+            // TODO(@jroesch): we really need to standaridize the bit width of
+            // all of the shape manipulating code.
+            CHECK_LE(tensor.dtype.bits, 64)
+              << "found " << tensor.dtype.bits;
+            int64_t* int_ptr = reinterpret_cast<int64_t*>(tensor.data);
```

Comment:
This is why test fails. For i386, we need to use int32_t. One work around would be checking the number of bits in `tensor.dtype.bits`. But we do need a better way to unify it.

Reply:
Okay I'm going to put a todo, and we should follow up with discussion about this this week, there are way too many places where dtypes are implicitly set.